File: src-tauri/src/commands/keymap.rs
//! Keymap IPC — list / set / unset / reset.

use std::sync::Arc;

use crate::services::keymap::{KeymapBinding, KeymapStore};
use crate::types::DuetError;

#[tauri::command]
#[specta::specta]
pub async fn keymap_list(
    store: tauri::State<'_, Arc<KeymapStore>>,
) -> Result<Vec<KeymapBinding>, DuetError> {
    Ok(store.inner().list().await)
}

#[tauri::command]
#[specta::specta]
pub async fn keymap_set(
    key: String,
    command_id: String,
    store: tauri::State<'_, Arc<KeymapStore>>,
) -> Result<Vec<KeymapBinding>, DuetError> {
    store.inner().set(key, command_id).await
}

#[tauri::command]
#[specta::specta]
pub async fn keymap_unset(
    key: String,
    store: tauri::State<'_, Arc<KeymapStore>>,
) -> Result<Vec<KeymapBinding>, DuetError> {
    store.inner().unset(&key).await
}

#[tauri::command]
#[specta::specta]
pub async fn keymap_reset(
    store: tauri::State<'_, Arc<KeymapStore>>,
) -> Result<Vec<KeymapBinding>, DuetError> {
    store.inner().reset().await
}

File: src/stores/keymap.ts
import { create } from "zustand";
import { commands } from "@/types/bindings";
import { useToast } from "@/stores/toast";
import { formatErr } from "@/lib/error";
import type { KeymapBinding } from "@/types/bindings";

interface State {
  bindings: KeymapBinding[];
  setAll: (b: KeymapBinding[]) => void;
}

export const useKeymap = create<State>((set) => ({
  bindings: [],
  setAll: (bindings) => set({ bindings }),
}));

/** override (있으면) 또는 defaultKey */
export function effectiveKey(
  commandId: string,
  bindings: KeymapBinding[],
  defaultKey?: string,
): string | undefined {
  const override = bindings.find((b) => b.command_id === commandId);
  return override?.key ?? defaultKey;
}

export async function bootstrapKeymap(): Promise<void> {
  const r = await commands.keymapList();
  if (r.status === "ok") useKeymap.getState().setAll(r.data);
}

export async function setKeymap(
  key: string,
  commandId: string,
): Promise<boolean> {
  const r = await commands.keymapSet(key, commandId);
  if (r.status === "ok") {
    useKeymap.getState().setAll(r.data);
    return true;
  }
  return false;
}

export async function unsetKeymap(key: string): Promise<void> {
  const r = await commands.keymapUnset(key);
  if (r.status === "ok") useKeymap.getState().setAll(r.data);
  else useToast.getState().show(`Unbind key: ${formatErr(r.error)}`, "error");
}

export async function resetKeymap(): Promise<void> {
  const r = await commands.keymapReset();
  if (r.status === "ok") useKeymap.getState().setAll(r.data);
  else useToast.getState().show(`Reset keymap: ${formatErr(r.error)}`, "error");
}
